cmd/main.go

```diff
@@ -5,160 +5,284 @@ import (
+	}
+
+	var path sphinx.PaymentPath
+	for i, hop := range data.Hops {
 		binKey, err := hex.DecodeString(hop.PublicKey)
-		if err != nil || len(binKey) != 33 {
-			log.Fatalf("%s is not a valid hex pubkey %s", hop.PublicKey, err)
+		if err != nil {
+			return nil, nil, nil, err
 		}
 
 		pubkey, err := btcec.ParsePubKey(binKey)
 		if err != nil {
-			log.Fatalf("%s is not a valid hex pubkey %s", hop.PublicKey, err)
+			return nil, nil, nil, err
 		}
 
 		path[i].NodePub = *pubkey
 
 		payload, err := hex.DecodeString(hop.Payload)
 		if err != nil {
-			log.Fatalf("%s is not a valid hex payload %s",
-				hop.Payload, err)
+			return nil, nil, nil, err
 		}
 
 		hopPayload, err := sphinx.NewHopPayload(nil, payload)
 		if err != nil {
-			log.Fatalf("unable to make payload: %v", err)
+			return nil, nil, nil, err
 		}
 
 		path[i].HopPayload = hopPayload
-
-		fmt.Fprintf(os.Stderr, "Node %d pubkey %x\n", i, pubkey.SerializeCompressed())
 	}
-	return &path, sessionKey, nil
+
+	return &path, sessionKey, assocData, nil
 }
 
-// main implements a simple command line utility that can be used in order to
-// either generate a fresh mix-header or decode and fully process an existing
-// one given a private key.
-func main() {
-	args := os.Args
+func generate(ctx *cli.Context) error {
+	file := ctx.String("file")
+	jsonSpec, err := os.ReadFile(file)
+	if err != nil {
+		return fmt.Errorf("unable to read JSON onion spec from file "+
+			"%v: %v", file, err)
+	}
 
-	assocData := bytes.Repeat([]byte{'B'}, 32)
+	var spec pathData
+	if err := json.Unmarshal(jsonSpec, &spec); err != nil {
+		return fmt.Errorf("unable to peel JSON onion spec: %v", err)
+	}
 
-	if len(args) < 3 {
-		fmt.Printf("Usage: %s (generate|decode) <input-file>\n", args[0])
-		return
-	} else if args[1] == "generate" {
-		var spec OnionSpec
+	path, sessionKey, assocData, err := parsePathData(spec)
+	if err != nil {
+		return fmt.Errorf("could not peel onion spec: %v", err)
+	}
 
-		jsonSpec, err := ioutil.ReadFile(args[2])
-		if err != nil {
-			log.Fatalf("Unable to read JSON onion spec from file %v: %v", args[2], err)
-		}
+	msg, err := sphinx.NewOnionPacket(
+		path, sessionKey, assocData, sphinx.DeterministicPacketFiller,
+	)
+	if err != nil {
+		return fmt.Errorf("error creating message: %v", err)
+	}
 
-		if err := json.Unmarshal(jsonSpec, &spec); err != nil {
-			log.Fatalf("Unable to parse JSON onion spec: %v", err)
-		}
+	w := bytes.NewBuffer([]byte{})
+	err = msg.Encode(w)
+	if err != nil {
+		return fmt.Errorf("error serializing message: %v", err)
+	}
 
-		path, sessionKey, err := parseOnionSpec(spec)
-		if err != nil {
-			log.Fatalf("could not parse onion spec: %v", err)
-		}
+	fmt.Printf("%x\n", w.Bytes())
+	return nil
+}
 
-		msg, err := sphinx.NewOnionPacket(
-			path, sessionKey, assocData,
-			sphinx.DeterministicPacketFiller,
-		)
-		if err != nil {
-			log.Fatalf("Error creating message: %v", err)
-		}
+type onionInfo struct {
+	SessionKey     string `json:"session_key"`
```

Comment:
`SessionKey` may be misleading here, isn't this the node priv key? For me the session key would be the ephemeral key E=eG (e or E) rather, please correct me if there's another interpretation :slightly_smiling_face: